src/util/LimitedAllocator.h

```diff
@@ -0,0 +1,164 @@
+//
+// Created by johannes on 27.04.20.
+//
+
+#ifndef QLEVER_LIMITEDALLOCATOR_H
+#define QLEVER_LIMITEDALLOCATOR_H
+
+#include <atomic>
+#include "Synchronized.h"
+
+namespace ad_utility {
+
+namespace detail {
+
+class LimitException : public std::exception {
+  const char* what() const noexcept override {
+    return "Tried to allocate more than the specified limit. Consider clearing "
+           "the cache or allowing more memory for QLever during startup";
+  }
+};
+
+// Basically holds an int that can be decreased by calls to allocate or
+// decreased by calls to deallocate. Trying to decrease below 0 will throw a
+// LimitException
+class AllocationLimits {
+  size_t free_;  // the number of free bytes
+                 // throw LimitException on failure
+ public:
+  AllocationLimits(size_t n) : free_(n) {}
+
+  void allocate(size_t n) {
+    if (n <= free_) {
+      free_ -= n;
+    } else {
+      throw LimitException{};
+    }
+  }
+
+  void deallocate(size_t n) { free_ += n; }
+  [[nodiscard]] size_t numFreeBytes() const { return free_; }
+};
+
+/*
+ * Wrapper around AllocationLimits that makes them shared
+ * Copying this class will share the same allocation limit.
+ * This class is threadsafe via ad_utilitty::Synchronized
+ */
+class AllocationState {
+ public:
+  AllocationState() = delete;
+  using T =
+      std::shared_ptr<ad_utility::Synchronized<AllocationLimits, SpinLock>>;
+  explicit AllocationState(T ptr) : ptr_{std::move(ptr)} {}
+  T& ptr() { return ptr_; }
+  const T& ptr() const { return ptr_; }
+
+  friend bool operator==(const AllocationState& a, const AllocationState& b) {
+    return a.ptr_ == b.ptr_;
+  }
+
+ private:
+  T ptr_;
+};
+}  // namespace detail
```

Comment:
The mere implementation details, that shouldn't be called from outside should be marked in some way. (Another aspect is, that users should not rely on a certain functionality of these implementation details, because they might change over time, whereas the "public" interface should not. 

At a second thought I would even consider making the name more explicit like `limited_allocator_detail' to not accidentally mix it with other `detail` namespace. (The outer namespage here is `ad_utility` which is quite widespread).